RedditSharp/WebAgent.cs

```diff
@@ -112,13 +66,38 @@ public int RequestsThisBurst
         public WebAgent( string accessToken ) {
             RootDomain = OAuthDomainUrl;
             AccessToken = accessToken;
+            RateLimitUsed = 0;
+            RateLimitReset = DateTimeOffset.UtcNow;
+            RateLimitRemaining = IsOAuth ? 60 : 30;
         }
 
         /// <inheritdoc />
-        public virtual async Task<JToken> ExecuteRequestAsync(HttpRequestMessage request)
+        public virtual async Task<JToken> ExecuteRequestAsync(Func<HttpRequestMessage> request)
         {
-            await EnforceRateLimit().ConfigureAwait(false);
-            var response = await _httpClient.SendAsync(request).ConfigureAwait(false);
+            if (request == null)
+              throw new ArgumentNullException(nameof(request));
+            const int maxTries = 20;
+            HttpResponseMessage response;
+            var tries = 0;
+            do {
+              await rateLimitLock.WaitAsync().ConfigureAwait(false);
```

Comment:
That is manageable.

Simply run the rate limit enforcement as it was before, but still keep track of the header values. If the header values state that no more requests can be made, even if the previously defined rate limit checks say they can, it will still wait until the reset period is over.